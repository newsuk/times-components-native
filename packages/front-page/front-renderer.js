import React from "react";
import coreRenderers from "@times-components-native/markup";
import { Text } from "react-native";

export const PARAGRAPH_INDENT = `\u3000`; // approximates a tab

export default {
  ...coreRenderers,
  link(key, attributes, renderedChildren) {
    return <Text key={key}>{renderedChildren}</Text>;
  },
  paragraph(key, attributes, renderedChildren) {
    return (
      <Text key={key}>
        {attributes.tab && PARAGRAPH_INDENT}
        {renderedChildren}
        {`\n`}
      </Text>
    );
  },
};
