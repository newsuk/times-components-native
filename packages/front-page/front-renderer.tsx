import React from "react";
import { StyleSheet, Text, TextStyle } from "react-native";
import coreRenderers from "@times-components-native/markup";

type Renderer = (
  key: string,
  attributes: { value: string; tab?: boolean },
  renderedChildren?: React.Component,
) => React.ComponentElement<any, any> | string;

interface Input {
  onParagraphTextLayout?: (event: any) => void;
  leftAligned?: boolean;
  renderOptions?: TextStyle;
}
type GetRenderers = (input: Input) => { [key: string]: Renderer };
export const PARAGRAPH_INDENT_CHAR = `\u3000`; // approximates a 2-space tab

export const getRenderers: GetRenderers = ({
  renderOptions,
  onParagraphTextLayout,
}) => ({
  ...coreRenderers,
  paragraph(key, attributes, renderedChildren) {
    return (
      <Text
        key={key}
        style={[!!renderOptions && renderOptions]}
        // @ts-ignore onTextLayout does exist on Text component
        onTextLayout={onParagraphTextLayout}
      >
        {attributes?.tab && PARAGRAPH_INDENT_CHAR}
        {renderedChildren}
      </Text>
    );
  },
  invisible(key, attributes) {
    return (
      <Text key={key} style={styles.invisible}>
        {attributes.value}
      </Text>
    );
  },
  link(key, _attributes, renderedChildren) {
    return <Text key={key}>{renderedChildren}</Text>;
  },
});

const styles = StyleSheet.create({
  invisible: {
    color: "rgba(0,0,0,0)",
  },
});
