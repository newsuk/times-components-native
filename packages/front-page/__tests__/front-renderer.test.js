import ReactTestRenderer from "react-test-renderer";

import { getRenderers } from "../front-renderer";

const renderedChildren = "Some Text";

const renderers = getRenderers({ renderOptions: { fontSize: 14 } });
describe("front renderers", () => {
  describe("paragraph renderer", () => {
    it("renders paragraph", () => {
      const paragraphRenderer = renderers.paragraph(
        "some-key",
        { value: "some value" },
        renderedChildren,
      );
      const renderer = ReactTestRenderer.create(paragraphRenderer);

      expect(renderer.toJSON()).toMatchSnapshot();
    });

    it("renders paragraph with tab", () => {
      const paragraphRenderer = renderers.paragraph(
        "some-key",
        { value: "some value", tab: true },
        renderedChildren,
      );
      const renderer = ReactTestRenderer.create(paragraphRenderer);

      expect(renderer.toJSON()).toMatchSnapshot();
    });

    it("renders paragraph with new line at end", () => {
      const renderersWithNewLine = getRenderers({ addNewLine: true });

      const paragraphRenderer = renderersWithNewLine.paragraph(
        "some-key",
        { value: "some value", tab: true },
        renderedChildren,
      );
      const renderer = ReactTestRenderer.create(paragraphRenderer);

      expect(renderer.toJSON()).toMatchSnapshot();
    });
  });

  describe("link text renderer", () => {
    it("renders link text", () => {
      const textRenderer = renderers.link("some-key", {}, renderedChildren);
      const renderer = ReactTestRenderer.create(textRenderer);

      expect(renderer.toJSON()).toMatchSnapshot();
    });
  });
});
