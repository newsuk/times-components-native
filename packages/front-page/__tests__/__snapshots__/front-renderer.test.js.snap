// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`front renderers link text renderer renders link text 1`] = `
<Text>
  Some Text
</Text>
`;

exports[`front renderers paragraph renderer renders paragraph 1`] = `
<Text
  style={
    Array [
      Object {
        "fontSize": 14,
      },
    ]
  }
>
  Some Text
</Text>
`;

exports[`front renderers paragraph renderer renders paragraph with new line at end 1`] = `
<Text
  style={
    Array [
      false,
    ]
  }
>
  　
  Some Text
  

</Text>
`;

exports[`front renderers paragraph renderer renders paragraph with tab 1`] = `
<Text
  style={
    Array [
      Object {
        "fontSize": 14,
      },
    ]
  }
>
  　
  Some Text
</Text>
`;
